### PR TITLE
feat(dashboards): MPL parameter values + building-dashboards skill overhaul

### PR DESCRIPTION
**Pull request by Licenser**
## Summary

Originally a focused docs change for MPL parameter values in `scripts/metrics-query`. While exercising the change against real Kubernetes / API-server dashboard rebuilds across seven agent sessions, the `building-dashboards` skill kept exposing footguns. The branch grew into a full overhaul of that skill — new tooling, two new reference docs, a restructured SKILL.md, and a tightened ruleset — alongside the original parameter-passing feature.

The diff: **2,840 insertions / 1,134 deletions across 25 files** in three skills (`building-dashboards`, `query-metrics`, `controlling-costs`).

---

### 1. MPL parameter values (the original feature)

Commits: `97e1e32`, `66ca029`.

- **`scripts/metrics-query`** (vendored in both `query-metrics` and `building-dashboards`): new repeatable `-p / --param name=value` flag. Each entry is forwarded under `params.param__<name>` on the JSON body of `POST /v1/query/_mpl`; the script applies the `param__` prefix, strips the leading `$`, and passes the value through verbatim. Dataset extraction now skips leading `param ...;` declarations so a colon inside a declaration doesn't shadow the `dataset:metric` separator.
- **`skills/query-metrics/SKILL.md`**: new "Passing parameter values" subsection documenting the API contract, request-body shape, CLI invocation, and required-vs-optional behavior. Defers MPL-side declaration syntax and per-type literal formatting to `metrics-spec`.
- **`skills/building-dashboards/reference/metrics-mpl.md`**: single pointer to the new subsection in `query-metrics`. No content duplication.
- **`scripts/metrics-info`**: now exposes metric `type` / `temporality` / `unit` metadata, used by the new `unit-for` helper and consumers downstream.

API contract:

```json
{
  "apl": "param $svc: string; param $window: Duration; `otel-metrics`:`http.server.duration` | where `service.name` == $svc | align to $window using avg",
  "startTime": "now-1h",
  "endTime": "now",
  "params": {
    "param__svc": "\"frontend\"",
    "param__window": "5m"
  }
}
```

CLI:

```bash
scripts/metrics-query \
  -p svc='"frontend"' \
  -p window='5m' \
  prod \
  'param $svc: string; param $window: Duration; `otel-metrics`:`http.server.duration` | where `service.name` == $svc | align to $window using avg' \
  now-1h now
```

The new flag is **non-breaking** — when absent, the request body is unchanged from before, and existing four-positional-arg invocations continue to work.

### 2. New `building-dashboards` tooling

Commit: `fc516bc` (with `dashboard-validate` and `resolve-url` env support added in `080f2ce`).

| Script                                          | Purpose                                                                                                                |
| :---------------------------------------------- | :--------------------------------------------------------------------------------------------------------------------- |
| `scripts/chart-add`                             | Append a chart to a dashboard JSON, with consistent panel-id assignment and validation hooks.                          |
| `scripts/dashboard-assemble`                    | Build a complete dashboard from a manifest of charts + layout, replacing ad-hoc hand-assembly of dashboard JSON.       |
| `scripts/layout-pack`                           | Compute a layout grid from chart metadata (size hints, groupings) so authors stop placing panels by hand.              |
| `scripts/dashboard-validate`                    | Validate a dashboard JSON against the API before deploy. Now respects `AXIOM_URL_OVERRIDE` (mirrors `axiom-api`).      |
| `scripts/metrics/mpl-validate-chart`            | Validate a single MPL chart definition against the live spec before it ever reaches the dashboard.                     |
| `scripts/metrics/unit-for`                      | Helper that resolves the correct `unit` / `customUnits` chart fields from metric metadata (incl. `%`-fraction handling). |
| `scripts/metrics/resolve-url`                   | Now honors `AXIOM_URL_OVERRIDE` so a single env var works across `axiom-api`, `dashboard-validate`, and the metrics scripts. |
| `tests/unit-for.sh`                             | Test suite for the unit-for helper, covering OTel fraction → percent conversion and the `%` edge case.                 |

### 3. New reference documents

Commit: `080f2ce`.

- **`reference/grafana-migration.md`** — canonical-spec projection (jq), top-level field mapping, visualization-type table, name mapping (Prom → OTel rename rules), reverse-tag discovery for missing labels, and a migration checklist.
- **`reference/promql-to-mpl.md`** — mechanical translation of PromQL selectors, groupings, `rate`, histograms, ratios, boolean step functions, and reverse-tag discovery. Includes the rule "selector values not in the dataset are aliases — cite the source."

### 4. Skill rules added or tightened

Across `c8b064b`, `24f87c6`, `acdb6c3`, `080f2ce`, `fc516bc`:

- **Compute What's Asked, or Defer With a Blocker.** If a panel can't be computed, replace it with a `Note` documenting the blocker. Never silently substitute a different quantity.
- **Required Chart Unit Configuration matrix.** Per-chart-type unit rules covering the asymmetry (e.g., `customUnits` is effectively `Statistic`-only; `Note` panel `text` / `variant` live at chart top-level, not under `options`).
- **OTel percentages arrive as 0–1 fractions.** Must be `| map * 100` before being rendered with `Percent100`.
- **No inline time ranges in chart queries.** Per-panel time override is UI-only; `overrideDashboardTimeRange` is silently dropped on data charts and rejected on Note. Documents the `$__interval` test pattern.
- **`align` belongs in `compute`**, not applied afterward (already shipped via PR #48 / `4517dfa`; reinforced here).
- **Unit-handling guidance** with the `unit-for` helper (commits `c8b064b`, `24f87c6`).

### 5. SKILL.md restructure

`skills/building-dashboards/SKILL.md` was tightened end-to-end. New top-level sections (some renamed, some new):

```
Philosophy
Entry Points                  ← now includes Grafana entry path
Intake: What to Ask First
Dashboard Blueprint
Required Chart Structure
Chart Unit Configuration      ← new, tied to the unit matrix above
Compute or Defer              ← new, tied to the defer-with-blocker rule
Chart Types
Chart Configuration
APL Patterns
Layout Composition
Dashboard Settings
Setup
Deployment
Sibling Skill Integration
Templates
Common Pitfalls
Reference
```

`skills/query-metrics/SKILL.md` was similarly restructured (sections: Workflow, Choosing a Query Shape, Query Metrics, Discovery, Error Handling, Scripts) and lost ~325 lines of redundant prose.

### 6. Cross-skill sync

Helper scripts that exist in both `query-metrics/scripts/` and `building-dashboards/scripts/metrics/` are kept byte-identical per the rule in `metrics-mpl.md`. This branch synchronizes both copies of:

- `metrics-info` (added type/temporality/unit metadata)
- `metrics-query` (added `-p / --param`)
- `resolve-url` (added `AXIOM_URL_OVERRIDE`)

Also updated: `skills/controlling-costs/reference/dashboard-panels.md`, `skills/controlling-costs/templates/dashboard.json`, and `skills/building-dashboards/templates/org-usage-cost-control.json` (extended to use the new tooling and rules).

---

## Verification

The MPL parameter feature was verified against the live MPL spec (`scripts/metrics-spec axiom ha`) and a live deployment:

| Claim                                             | Evidence                                                                                                                                                      |
| :------------------------------------------------ | :------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Key = `param__<name>` with `$` stripped           | Spec §Query Parameters: `$__interval`→`param____interval`, `$ds`→`param__ds`, `$re`→`param__re`, `$str`→`param__str`                                          |
| Value = verbatim MPL literal                      | Live test passing `-p eid='"sensor.heizstrom_power"'` returned series identical to a non-parameterized control with the literal substituted directly          |
| JSON `params` body works                          | End-to-end test against `axiom` deployment returned identical results to the literal control                                                                  |
| Required parameters must be supplied              | Live test omitting a required param returned HTTP 400: `"The following params were declared but not provided: eid"`                                           |
| Type-name capitalization                          | Spec uses lowercase `string` and capitalized `Duration` / `Dataset` / `Regex`; examples follow this convention                                                |

The dashboard-skill changes were verified by running the standard Grafana **Kubernetes / API server** dashboard rebuild against `k8s-metrics-staging` across seven agent sessions (see retrospective below).

### Notes

- The spec's optional-parameter mechanism is `Option<T>` + `ifdef`. Docs describe the observable behavior (optional may be omitted, required is required) and defer syntax to `metrics-spec`.
- `Option<T>` was rejected by the parser on the `axiom` deployment build I tested ("expected param type"), even though the spec doesn't flag it as unimplemented. The docs in this PR don't depend on that feature being live.

---

## Session retrospective — Kubernetes / API server dashboard iterations

The dashboards-skill changes on this branch were exercised by re-deploying the standard Grafana **Kubernetes / API server** dashboard against `k8s-metrics-staging` across seven agent sessions. Each session ended with the skill being tightened based on what broke; v6 was a one-shot success against the post-v5 skill.

| #   | Duration   | User msgs (dashboard / follow-up) | Assistant turns | Tool calls | Tool-call mix                          | Tools/min | Outcome                                        |
| :-- | ---------: | :-------------------------------- | --------------: | ---------: | :------------------------------------- | --------: | :--------------------------------------------- |
| v0a | 3 m 44 s   | 2 / 0                             |              28 |         28 | bash 26, read 2                         |       7.5 | abandoned mid-discovery                         |
| v0b | 1 h 47 m   | 3 / 4                             |              99 |         95 | bash 84, read 5, write 4, edit 2        |       0.9 | success + spawned `axiom-staging` skill         |
| v2  | 36 m 39 s  | 1 / 1                             |              62 |         60 | bash 50, edit 5, read 4, write 1        |       1.6 | success                                         |
| v3  | 13 m 58 s  | 2 / 1                             |              56 |         53 | bash 45, read 5, edit 2, write 1        |       3.8 | success                                         |
| v4  | 31 m 40 s  | 1 / 2 (+1 meta)                   |             111 |        107 | bash 80, edit 11, read 10, write 6      |       3.4 | success                                         |
| v5  | 42 m 30 s  | 2 / 6                             |              92 |         85 | bash 71, read 11, write 2, edit 1       |       2.0 | success (after fixing inline-time bug)          |
| v6  | 6 m 24 s   | 1 / 0                             |              55 |         59 | bash 49, read 5, ls 3, write 1, edit 1  |       9.2 | one-shot success                                |

### What each session contributed back to the skills

- **v0b** — created `~/.agents/skills/axiom-staging/` capturing the staging edge URL override (`us-east-1.edge.staging.axiomtestlabs.co`) so later sessions stop reinventing it. (Generalized later in this PR as `AXIOM_URL_OVERRIDE` in `resolve-url` / `dashboard-validate`.)
- **v2** — `building-dashboards`: noted that OTel percentages arrive as 0–1 fractions and must be `| map * 100` before being rendered with `Percent100` (commits `c8b064b`, `24f87c6`).
- **v3** — `building-dashboards`: rule that `align` must be pulled into `compute`, not applied afterwards (already shipped in PR #48 as `4517dfa`, separate from this PR).
- **v4** — `building-dashboards`: documented that `customUnits` is effectively `Statistic`-only and that `Note` panel `text` / `variant` live at chart top-level, not under `options` (commits `c8b064b`, `24f87c6`).
- **v5** — `building-dashboards`: forbade inline time ranges in chart queries, documented `$__interval` test pattern (commit `acdb6c3`).
- **v6** — no skill change required; one-shot success against the post-v5 skill, validating the rules added in v2/v4/v5.

The Grafana migration guide (`grafana-migration.md`) and PromQL→MPL translation rules (`promql-to-mpl.md`) introduced in `080f2ce`, plus the assembly tooling (`chart-add` / `dashboard-assemble` / `layout-pack` / `mpl-validate-chart`) in `fc516bc`, are the **codified output** of these seven sessions — turning hand-assembly + ad-hoc validation into a repeatable workflow.

### Read of the table

- **v6 is dramatically more efficient** than every prior run: 1 prompt, 6 minutes, 55 turns, 59 tool calls — and the highest tool throughput (9.2 / min). The previous fastest success (v3) needed 56 turns over 14 minutes.
- **v0a's high tools/min is misleading** — it's a short, truncated session that never wrote a dashboard.
- **v0b** is the heaviest absolute (99 turns, 95 tool calls, ~108 minutes) because half of it was building a brand-new skill, not the dashboard.
- **v4 is the chattiest** (111 turns, 107 tool calls) because it iterated heavily on the `building-dashboards` skill while debugging custom-units.
- `bash` dominates everywhere (~80–90% of all tool calls).
- v6 has more tool calls than assistant turns (59 > 55) — some turns batched 2+ tool calls in one message, an efficiency pattern the matured skills now encourage.

**Trend:** real user messages dropped from v5 → v6 (8 → 1), the cleanest signal that the v5 skill changes stuck — v6 needed zero corrective follow-ups.

### Scope note

The original PR scope (`97e1e32`, `66ca029`) was the MPL `params` feature alone, from a separate session on May 7. The seven sessions above drove the dashboard-skill commits (`c8b064b`, `24f87c6`, `acdb6c3`, `080f2ce`, `fc516bc`) that this PR now also carries.

### CODE REVIEW

**Review by aisha331 — APPROVED** [review]

File: skills/building-dashboards/scripts/dashboard-assemble [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+# `dashboard-validate` and `dashboard-create`.
+#
+# Examples:
+#   chart-add --type Note --id intro --name Notice --text "..." \
+#     > /tmp/d/c-intro.json
+#   chart-add --type Statistic --id avail --name "Availability" \
+#       --apl '...' --dataset k8s-metrics-staging --unit "%" \
+#     > /tmp/d/c-avail.json
+#   layout-pack intro:Note avail:Statistic > /tmp/d/layout.json
+#
```
Comment (aisha331) [inline skills/building-dashboards/scripts/dashboard-assemble:72]: shouldn’t this be --mpl for metrics?

File: skills/building-dashboards/scripts/layout-pack [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+#
+#   # Full dashboard skeleton
+#   layout-pack notice:Note \
+#               avail:Statistic budget:8x3 \
+#               read-avail:Statistic read-rps:TimeSeries read-err:TimeSeries read-lat:TimeSeries \
+#               cpu:TimeSeries mem:TimeSeries goroutines:TimeSeries
+#
+# Composing with chart-add:
+#   chart-add --type Note --id intro --name "..." --text "..."  > /tmp/c-intro.json
+#   chart-add --type Statistic --id avail --name "..." --apl '...' --dataset ds --unit "%" > /tmp/c-avail.json
```
Comment (aisha331) [inline skills/building-dashboards/scripts/layout-pack]: same here. https://github.com/axiomhq/skills/pull/49/changes#diff-319606d3a7f0aad809c97f0f991c47745c16a630dd1d6de62f4b384467ab0718R150-R158

File: skills/building-dashboards/scripts/dashboard-assemble (hunk 1 above)
Comment (Licenser) [reply skills/building-dashboards/scripts/dashboard-assemble:72]: good catch! Fixed

File: skills/building-dashboards/scripts/layout-pack (hunk 2 above)
Comment (Licenser) [reply skills/building-dashboards/scripts/layout-pack]: good catch! Fixed
